[https://nvbugs/5625990][fix] Add synchronization to make sure async actions are completed before each other

eopXD opened the pull request:
NOTE: This MR is currently based and depends on https://github.com/NVIDIA/TensorRT-LLM/pull/9056. This MR will be rebased and re-organized after https://github.com/NVIDIA/TensorRT-LLM/pull/9056 lands.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved synchronization between GPU memory transfer operations and buffer management to prevent race conditions during cache operations.

* **Refactor**
  * Enhanced KV cache management with refined block reuse tracking and GPU-to-GPU copy operations for optimized memory transfer performance.
  * Strengthened resource cleanup and synchronization during model warmup initialization.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

Add synchronization to make sure async actions are completed before each other

We want onboard/offload actions to be completed before
kernel(mBufferManager) performs and vice versa too.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

This fixes the incorrect partial block copy functionality exposed
through test case `accuracy/test_llm_api_pytorch.py::TestGemma3_1BInstruct::test_auto_dtype_vswa_reuse_low_memory_available_partial_reuse`


## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

thorjohnsen committed:
Consider block granularity when computing remaining tokens

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
precommit run

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_bug_5375590

thorjohnsen committed:
Remove test waiver

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'user/tjohnsen/fix_bug_5375590' of github.com:thorjohnsen/TensorRT-LLM into user/tjohnsen/fix_bug_5375590
.

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_bug_5375590

thorjohnsen committed:
Remove test waiver

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_bug_5375590

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_bug_5375590

Tabrizian committed:
Add refresh blocks

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

.

thorjohnsen committed:
Fix transfer manager synchronization issues

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

.

thorjohnsen committed:
Fix merge issues

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Bug fix

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Another fix

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

thorjohnsen committed:
precommit run

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'user/tjohnsen/fix_5627710' of github.com:thorjohnsen/TensorRT-LLM into user/tjohnsen/fix_5627710
.

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

thorjohnsen committed:
Fix broken pybind

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'user/tjohnsen/fix_5627710' of github.com:thorjohnsen/TensorRT-LLM into user/tjohnsen/fix_5627710
.

thorjohnsen committed:
Move refreshBlocks call to account for addToken calls

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
precommit run

Signed-off-by: thorjohnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

thorjohnsen committed:
Update cpp/tensorrt_llm/batch_manager/kvCacheTransferManager.cpp

Co-authored-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>
Signed-off-by: Thor Johnsen <41591019+thorjohnsen@users.noreply.github.com>

thorjohnsen committed:
Merge branch 'main' into user/tjohnsen/fix_5627710

eopXD committed:
Merge remote-tracking branch 'thor/user/tjohnsen/fix_bug_5375590' into try-thor-fix-transfer-manager

eopXD committed:
Merge remote-tracking branch 'thor/user/tjohnsen/fix_5627710' into try-thor-fix-transfer-manager

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheTransferManager.cpp:
```diff
@@ -207,47 +220,156 @@ void KVCacheTransferManager::copyBlock(BlockPtr const& src, BlockPtr const& dst,
+    syncPendingOnboardTransfers(offloadedBlock, block);
+    copyBlock(offloadedBlock, block, pools, false, numTokensToCopy, mode, directory);
+    recordPendingOnboardTransfers(block);
+}
```
Nice catch

nvpohanh commented on cpp/include/tensorrt_llm/batch_manager/kvCacheTransferManager.h:
```diff
@@ -71,12 +98,23 @@ class KVCacheTransferManager
         std::string const& directory = "");
 
+    /*!
+     * \brief Directly copy a block from gpu to gpu.
```
Is this asynchronous from CPU's perspective? If so, we should mention that here. Something like:

```
\note The copy is launched on an internal stream asynchronously and this function may return before the copy is complete.
```

eopXD replied:
My approach has changed, a synchronize to mOnboardManager and mOffloadManager shall do. So the function is no longer needed.

eopXD committed:
[https://nvbugs/5625990][fix] Add synchronization to make sure async actions are completed before each other

We want onboard/offload actions to be completed before
kernel(mBufferManager) performs and vice versa too.

Signed-off-by: eopXD <yuehtingc@nvidia.com>

eopXD commented on cpp/tensorrt_llm/batch_manager/kvCacheTransferManager.cpp:
```diff
@@ -207,47 +220,156 @@ void KVCacheTransferManager::copyBlock(BlockPtr const& src, BlockPtr const& dst,
+{
+    syncPendingOnboardTransfers(offloadedBlock, block);
+    copyGPUtoGPU(offloadedBlock, block, pools);
+    recordPendingOnboardTransfers(offloadedBlock, block);
```
Approach has changed. Closing this one.
